File: python-pr/52/python/code-quality-readability-maintainability/password_checker_40.py
# {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
def assess_password_strength(password):
    if len(password) < 8:
        return 'weak'
    
    strength_criteria = {
        'has_uppercase': any(c.isupper() for c in password),
        'has_lowercase': any(c.islower() for c in password),
        'has_digit': any(c.isdigit() for c in password),
        'has_special': any(c in '!@#$%^&*()_+-=[]{}|;:,.<>?' for c in password)
    }
    
    criteria_met = sum(strength_criteria.values())
    
    if criteria_met >= 3:
        return 'strong'
    elif criteria_met >= 2:
        return 'medium'
    else:
        return 'weak'
# {/fact}

File: js-pr/56/javascript/code-quality-readability-maintainability/password_checker_40.js
// {fact rule=code-quality-readability-maintainability@v1.0 defects=0}
function assessPasswordStrength(password) {
    if (password.length < 8) {
        return 'weak';
    }
    
    const strengthCriteria = {
        hasUppercase: /[A-Z]/.test(password),
        hasLowercase: /[a-z]/.test(password),
        hasDigit: /\d/.test(password),
        hasSpecialChar: /[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/.test(password)
    };
    
    const criteriaMet = Object.values(strengthCriteria).filter(Boolean).length;
    
    if (criteriaMet >= 3) {
        return 'strong';
    } else if (criteriaMet >= 2) {
        return 'medium';
    } else {
        return 'weak';
    }
}
// {/fact}